src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -288,6 +311,41 @@ private String resolveSubject(
         return subject;
     }
 
+    private long resolveAdvancedNotificationBuffer(
+        JsonWebSignature verified,
+        int capabilities,
+        long expiresAtMillis)
+    {
+        long bufferDelta = 0;
+
+        if (hasChallengeCapability(capabilities))
+        {
+            try
+            {
+                if (verified != null)
+                {
+                    final JwtClaims claims = JwtClaims.parse(verified.getPayload());
+                    final NumericDate buffer = claims.getNumericDateClaimValue(CHALLENGE_AFTER_CLAIM);
+                    if (buffer != null)
+                    {
+                        bufferDelta = expiresAtMillis - buffer.getValueInMillis();
+                    }
+                }
+            }
+            catch (InvalidJwtException | JoseException | MalformedClaimException e)
+            {
+                // invalid token
+            }
+        }
+        return bufferDelta;
+    }
+
+    private boolean hasChallengeCapability(
+        int capabilities)
+    {
+        return (capabilities & CHALLENGE_MASK) == CHALLENGE_BIT;
+    }
```

Comment:
Hoist this into a class called `Capabilities` with a static method called `canChallenge` that has the same implementation as above.